internal/service/componentdescriptor/service.go

```diff
@@ -0,0 +1,223 @@
+	PullLayer(ctx context.Context, name, tag, digest string) (containerregistryv1.Layer, error)
+}
+
+type Service struct {
+	ociRepository OCIRepository
+}
+
+func NewService(ociRepository OCIRepository) (*Service, error) {
+	if ociRepository == nil {
+		return nil, fmt.Errorf("ociRepository must not be nil: %w", ErrInvalidArg)
+	}
+
+	return &Service{
+		ociRepository: ociRepository,
+	}, nil
+}
+
+func commonErrMsg(ocmi ocmidentity.Component) string {
+	return fmt.Sprintf("ocm artifact with name=%q and version=%q",
+		ocmi.Name(), ocmi.Version())
+}
+
+func (s *Service) GetComponentDescriptor(ctx context.Context, ocmi ocmidentity.Component) (*types.Descriptor, error) {
+	// Fetch the image config to get the ComponentDescriptor layer info
+	configBytes, err := s.ociRepository.GetConfigFile(ctx, ocmi.Name(), ocmi.Version())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get config file for %s: %w", commonErrMsg(ocmi), err)
+	}
+
+	ocmArtifactConfig := genericocireg.ComponentDescriptorConfig{}
+	err = json.Unmarshal(configBytes, &ocmArtifactConfig)
+	if err != nil {
+		return nil,
+			fmt.Errorf("failed to unmarshal config data into ComponentDescriptorConfig for %s: %w",
+				commonErrMsg(ocmi), err)
+	}
+
+	if ocmArtifactConfig.ComponentDescriptorLayer == nil {
+		return nil, fmt.Errorf("%w for %s", ErrLayerNil, commonErrMsg(ocmi))
+	}
+
+	compDescLayerDigest := ocmArtifactConfig.ComponentDescriptorLayer.Digest
+	if string(compDescLayerDigest) == "" {
+		return nil,
+			fmt.Errorf("%w for %s", ErrLayerDigestEmpty, commonErrMsg(ocmi))
+	}
+
+	layer, err := s.ociRepository.PullLayer(ctx, ocmi.Name(), ocmi.Version(), string(compDescLayerDigest))
+	if err != nil {
+		return nil, fmt.Errorf("failed to pull layer for ocm artifact with name=%q, version=%q and digest=%q: %w",
+			ocmi.Name(), ocmi.Version(), string(compDescLayerDigest), err)
+	}
+
+	ioh := &defaultExtractFileIOHelper{readAllFunc: io.ReadAll}
+	compdescBytes, err := extractFile(ioh, layer, ComponentDescriptorFileName)
+	if err != nil {
+		return nil,
+			fmt.Errorf("failed to extract component descriptor from layer fetched from %s with digest=%q: %w",
+				commonErrMsg(ocmi), string(compDescLayerDigest), err)
+	}
+
+	descriptor, err := deserialize(compdescBytes, ocmi)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.Descriptor{
+		ComponentDescriptor: descriptor,
+	}, nil
+}
+
+// deserialize decodes the component descriptor from its serialized form.
+func deserialize(compdescBytes []byte, ocmi ocmidentity.Component) (*compdesc.ComponentDescriptor, error) {
+	desc, err := compdesc.Decode(compdescBytes)
+	if err != nil {
+		return nil, fmt.Errorf("%w fetched from %s: %w",
+			ErrDecode, commonErrMsg(ocmi), err)
+	}
+	return desc, nil
+}
+
+// SimplifiedLayer represents only necessary part of containerregistryv1.Layer to simplify testing.
+type SimplifiedLayer interface {
+	Digest() (containerregistryv1.Hash, error)
+	Uncompressed() (io.ReadCloser, error)
+}
```

Comment:
This should not be needed. We can have `func extractFile(ioh extractFileIOHelper, layer containerregistryv1.Layer, fileName string) ([]byte, error) {` with mock layer

```go
type mockLayer struct {
	containerregistryv1.Layer

	errOnUncompressed error
	errOnDigest       error
}
```

Needed Uncompressed() and Digest() are overriden. Rest are implicitly exposed but will not work since we didn't provide an actual Layer